### PR TITLE
851: Implement Accounts & Relationships framework

### PR DESCRIPTION
**Pull request by raman-bt**

### CODE REVIEW

**Comment by bcarrier** [comment]
Can you fix the codacy issues?

File: bindings/java/src/org/sleuthkit/datamodel/Account.java [hunk 1]
```diff
@@ -18,23 +18,182 @@
...
+	public void addAttribute(BlackboardAttribute bbatr) throws TskCoreException {
+		getArtifact().addAttribute(bbatr);
+	}
+	
+	public long getArtifactId() {
+		return this.artifactId;
+	}
+	
+	private BlackboardArtifact getArtifact() throws TskCoreException {
+		if (null == this.artifact) {
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/Account.java]: What is the point of this "lazy loading" if this method is called in the constructors?

File: bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+	
+	Set<UnorderedPair<Long>> relationships = listToUnorderedPairs(accountIDs);
+	
+	Iterator<UnorderedPair<Long>> iter = relationships.iterator();
+	
+	while (iter.hasNext()) {
+		
+		try {
+			UnorderedPair<Long> accountPair = iter.next();
+			if (accountPair.first.longValue() != accountPair.second.longValue()) {// dont add a relationship with self
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java]: What if a relationship with self is what is intended such as an email to oneself?  Gmail jsut let me send an email to myself.

File: bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+	for (int i = 0; i < accountIDs.size(); i++ ) {
+		for (int j = i+1; j < accountIDs.size(); j++ ) {
+			relationships.add(new UnorderedPair<Long> (accountIDs.get(i), accountIDs.get(j) ));
+		}
+	}
+	
+	return relationships;
+}
+
+public final class UnorderedPair<T> {
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/CommunicationsManager.java]: Do we really need this class for such a lmited usecase?  Also it should probably be private.

File: bindings/java/src/org/sleuthkit/datamodel/SleuthkitCase.java [hunk 4]
```diff
@@ -201,6 +276,18 @@ private void init() throws Exception {
 	}
 
 	/**
+	 * Returns an instance of CommunicationsManager
+	 *
+	 * @return CommunicationsManager 
+	 */
+	public CommunicationsManager getCommunicationsManager() {
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/SleuthkitCase.java]: this should be synchronized for thread safety, shouldn't it?

File: bindings/java/src/org/sleuthkit/datamodel/MessageFolder.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	
+	public long getSrcOBjID() {
+		return this.srcObjID;
+	}
+	
+	public boolean hasSubfolders() {
+		return this.hasSubfolders;
+	}
+	
+	public void setHasSubfolders(boolean hasSubFolders) {
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/MessageFolder.java]: does this need to be thread safe?

File: bindings/java/src/org/sleuthkit/datamodel/SleuthkitCase.java [hunk 6]
```diff
@@ -2397,6 +2515,658 @@ public BlackboardArtifact getBlackboardArtifact(long artifactID) throws TskCoreE
...
+		
+		// Create a new artifact.
+		accountArtifact = newBlackboardArtifact(artifactType.getTypeID(), sourceObj.getId()); 
+		
+		Collection<BlackboardAttribute> attributes = new ArrayList<BlackboardAttribute>();
+		attributes.add(new BlackboardAttribute(ATTRIBUTE_TYPE.TSK_ACCOUNT_TYPE, moduleName, accountType.getTypeName()));
+		attributes.add(new BlackboardAttribute(ATTRIBUTE_TYPE.TSK_ID, moduleName, accountID));
+		accountArtifact.addAttributes(attributes);
+		
+		accountArtifact = getAccountArtifact(artifactType, accountType, accountID );
```
Comment (millmanorama) [inline bindings/java/src/org/sleuthkit/datamodel/SleuthkitCase.java]: do we need to go back to the DB here?

**Comment by raman-bt** [comment]
Jonathan - addressed your comments ( most -  I left the UnorderedPair class in place)

File: tsk/auto/db_postgresql.cpp [hunk 7]
```diff
@@ -607,8 +607,16 @@ int TskDbPostgreSQL::initialize() {
         "Error creating ingest_job_modules table: %s\n")
         ||
         attempt_exec
-        ("CREATE TABLE reports (report_id BIGSERIAL PRIMARY KEY, path TEXT NOT NULL, crtime INTEGER NOT NULL, src_module_name TEXT NOT NULL, report_name TEXT NOT NULL)","Error creating reports table: %s\n")) {
-            return 1;
+        ("CREATE TABLE reports (report_id BIGSERIAL PRIMARY KEY, path TEXT NOT NULL, crtime INTEGER NOT NULL, src_module_name TEXT NOT NULL, report_name TEXT NOT NULL)","Error creating reports table: %s\n")
+		||
+		attempt_exec
+		("CREATE TABLE account_types (account_type_id INTEGER PRIMARY KEY, type_name TEXT NOT NULL, display_name TEXT NOT NULL)", 
+		"Error creating reports table: %s\n")
```
Comment (millmanorama) [inline tsk/auto/db_postgresql.cpp]: the error string has "reports" instead of "account_types"

**Pull request merged**